engine/src/main/java/org/terasology/world/chunks/ChunkLightMergingSystem.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.world.chunks;
+
+import com.google.common.collect.Sets;
+import org.joml.Vector3ic;
+import org.terasology.engine.subsystem.common.ThreadManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.registry.In;
+import org.terasology.world.block.BlockRegion;
+import org.terasology.world.chunks.event.BeforeChunkUnload;
+import org.terasology.world.chunks.event.OnChunkLoaded;
+import org.terasology.world.propagation.light.LightMerger;
+
+import java.util.Set;
+
+/**
+ * System which will process light merging after chunk is ready.
+ */
+@RegisterSystem
+public class ChunkLightMergingSystem extends BaseComponentSystem {
+
+    private final Set<Vector3ic> loadedChunkPos = Sets.newConcurrentHashSet();
+    private final Set<Vector3ic> lightMergedChunks = Sets.newConcurrentHashSet();
+
+    @In
+    ChunkProvider chunkProvider;
+    @In
+    ThreadManager threadManager;
+
+    @ReceiveEvent
+    public void onChunkLoad(OnChunkLoaded chunkLoaded, EntityRef world) {
+        loadedChunkPos.add(chunkLoaded.getChunkPos());
+        processNearbyChunks(chunkLoaded.getChunkPos());
+    }
+
+    private void processNearbyChunks(Vector3ic chunkPos) {
+        for (Vector3ic candidateForLightMerging : new BlockRegion(chunkPos).expand(1, 1, 1)) {
+            if (lightMergedChunks.contains(candidateForLightMerging)) {
+                continue;
+            }
+            processCandidateChunk(candidateForLightMerging);
+            lightMergedChunks.add(candidateForLightMerging);
```

Comment:
Need to create a copy of the vector here, because it will be reused in the next iteration of the loop :wink: 
```suggestion
            lightMergedChunks.add(new Vector3i(candidateForLightMerging));
```